### PR TITLE
Initial VLAN implementation

### PR DESCRIPTION
**Pull request by guoshiuan**

### CODE REVIEW

File: dataplane/dplanerc/interface.go [hunk 1]
```diff
@@ -39,6 +39,7 @@ import (
 
 	log "github.com/golang/glog"
 
+	"github.com/openconfig/lemming/dataplane/proto/sai"
```
Comment (DanG100) [inline dataplane/dplanerc/interface.go:42]: this is already imported

File: dataplane/saiserver/routing.go [hunk 2]
```diff
@@ -630,6 +631,8 @@ func (ri *routerInterface) CreateRouterInterface(ctx context.Context, req *saipb
 		return nil, err
 	}
 
+	// Here we associate a port to its output interface. How is output port assigned?
```
Comment (DanG100) [inline dataplane/saiserver/routing.go:634]: when we create an interface we associate the interface to a port
we do this twice in two places: on to map in port to l3 interface and one on map to FIB output interface to output port

File: integration_tests/dataplane/mymac/mymac_test.go [hunk 3]
```diff
@@ -124,6 +124,16 @@ func configureDUT(t testing.TB, conn *grpc.ClientConn, dut *ondatra.DUTDevice) {
 	if err != nil {
 		t.Fatal(err)
 	}
+	// Add this port to default VLAN.
+	vc := saipb.NewVlanClient(conn)
+	ctx := context.Background()
+	if _, err := vc.CreateVlanMember(ctx, &saipb.CreateVlanMemberRequest{
+		VlanId:          proto.Uint64(4095), // the default VLAN ID.
```
Comment (DanG100) [inline integration_tests/dataplane/mymac/mymac_test.go:131]: shouldn't this be the OID of the VLAN object? not the vlan id

File: dataplane/saiserver/switch.go [hunk 4]
```diff
@@ -101,6 +101,8 @@ const (
 	hostifToPortTable     = "cpu-input"
 	portToHostifTable     = "cpu-output"
 	tunTermTable          = "tun-term"
+	VlanTable             = "vlan"
+	DefaultVlanId         = 4095 // ID of the default VLAN.
```
Comment (DanG100) [inline dataplane/saiserver/switch.go:105]: where did value come from?

File: dataplane/dplanerc/interface.go (hunk 1 above)
Comment (guoshiuan) [reply dataplane/dplanerc/interface.go]: Fixed

File: integration_tests/dataplane/mymac/mymac_test.go (hunk 3 above)
Comment (guoshiuan) [reply integration_tests/dataplane/mymac/mymac_test.go]: Fixed

File: dataplane/saiserver/switch.go (hunk 4 above)
Comment (guoshiuan) [reply dataplane/saiserver/switch.go]: Update comment.

File: dataplane/saiserver/routing.go (hunk 2 above)
Comment (guoshiuan) [reply dataplane/saiserver/routing.go]: Ack.

File: dataplane/dplanerc/interface.go [hunk 5]
```diff
@@ -940,6 +942,21 @@ func (ni *Reconciler) setupPorts(ctx context.Context) error {
 		}
 		data.portNID = nid.Nid
 
+		// Add this port to default VLAN.
+		swAttr, err := ni.switchClient.GetSwitchAttribute(ctx, &saipb.GetSwitchAttributeRequest{
```
[the remark this answers is not in the document]
Comment (guoshiuan) [reply dataplane/dplanerc/interface.go]: Move the logic to the port creation.

File: dataplane/dplanerc/interface.go [hunk 6]
```diff
@@ -940,6 +942,21 @@ func (ni *Reconciler) setupPorts(ctx context.Context) error {
...
+			Oid:      ni.switchID,
+			AttrType: []saipb.SwitchAttr{saipb.SwitchAttr_SWITCH_ATTR_DEFAULT_VLAN_ID},
+		})
+		if err != nil {
+			return err
+		}
+		if _, err := ni.vlanClient.CreateVlanMember(ctx, &saipb.CreateVlanMemberRequest{
```
[the remark this answers is not in the document]
Comment (guoshiuan) [reply dataplane/dplanerc/interface.go]: Move the logic to the port creation.

File: dataplane/saiserver/routing.go [hunk 7]
```diff
@@ -793,11 +842,127 @@ func (vlan *vlan) CreateVlan(context.Context, *saipb.CreateVlanRequest) (*saipb.
...
+	vlanAttrResp := &saipb.GetVlanAttributeResponse{}
+	if err := vlan.mgr.PopulateAttributes(vlanAttrReq, vlanAttrResp); err != nil {
+		return nil, err
+	}
+	vlanAttrResp.GetAttr().MemberList = append(vlanAttrResp.GetAttr().MemberList, mOid)
+	vlan.mgr.StoreAttributes(vOid, vlanAttrResp.GetAttr())
+	vlan.mu.Lock()
+	vlan.vlans[vOid][mOid] = &vlanMember{Oid: mOid, PortID: r.GetBridgePortId(), Vid: vId, Mode: r.GetVlanTaggingMode()}
+	vlan.mu.Unlock()
+	if member != nil {
```
[the remark this answers is not in the document]
Comment (guoshiuan) [reply dataplane/saiserver/routing.go:910]: No. It just means that the port was under other VLAN.

**Pull request merged**